src/com/facebook/buck/go/GoTestDescription.java

```diff
@@ -0,0 +1,172 @@
+            Suppliers.ofInstance(ImmutableSortedSet.<BuildRule>of())
+        ),
+        sourceResolver,
+        testMainGenerator,
+        srcs,
+        packageName
+    );
+    resolver.addToIndex(generatedTestMain);
+    return generatedTestMain;
+  }
+
+  @Override
+  public <A extends Arg> BuildRule createBuildRule(
+      TargetGraph targetGraph,
+      BuildRuleParams params,
+      BuildRuleResolver resolver,
+      A args) {
+    BuildTarget libraryTarget =
+        BuildTarget.builder(params.getBuildTarget())
+            .addFlavors(ImmutableFlavor.of("test-library"))
+            .build();
+
+    Path defaultPackageName = params.getBuildTarget().getBasePath();
+    defaultPackageName = defaultPackageName.resolveSibling(
+        defaultPackageName.getFileName() + "_test");
+
+    GoLibrary lib = GoDescriptors.createGoLibraryRule(
+        params.copyWithBuildTarget(libraryTarget),
+        resolver,
+        goBuckConfig,
+        args.packageName.transform(GoDescriptors.TO_GO_PACKAGE_NAME).or(defaultPackageName),
+        args.srcs,
+        args.compilerFlags.or(ImmutableList.<String>of())
+    );
+    resolver.addToIndex(lib);
+
+    GoTestMain generatedTestMain = requireTestMainGenRule(
+        params, resolver, args.srcs, lib.getGoPackageName());
+    GoBinary testMain = GoDescriptors.createGoBinaryRule(
+        params.copyWithChanges(
+            BuildTarget.builder(params.getBuildTarget())
+                .addFlavors(ImmutableFlavor.of("test-main"))
+                .build(),
+            Suppliers.ofInstance(ImmutableSortedSet.<BuildRule>of(lib)),
+            Suppliers.ofInstance(ImmutableSortedSet.<BuildRule>of(generatedTestMain))
+        ),
+        resolver,
+        goBuckConfig,
+        cxxPlatform,
+        ImmutableSet.<SourcePath>of(new BuildTargetSourcePath(generatedTestMain.getBuildTarget())),
+        args.compilerFlags.or(ImmutableList.<String>of()),
+        args.linkerFlags.or(ImmutableList.<String>of()));
+    resolver.addToIndex(testMain);
+
+    return new GoTest(
+        params.copyWithDeps(
+            Suppliers.ofInstance(ImmutableSortedSet.<BuildRule>of(testMain)),
+            Suppliers.ofInstance(ImmutableSortedSet.<BuildRule>of())
+        ),
+        new SourcePathResolver(resolver),
+        testMain,
+        args.labels.get(),
+        args.contacts.get(),
+        resolver.getAllRules(args.sourceUnderTest.get()),
+        args.runTestSeparately.or(false));
+  }
+
+  @SuppressFieldNotInitialized
+  public static class Arg implements HasSourceUnderTest {
+    public ImmutableSet<SourcePath> srcs;
+    public Optional<String> packageName;
+    public Optional<List<String>> compilerFlags;
+    public Optional<List<String>> linkerFlags;
+    public Optional<ImmutableSortedSet<BuildTarget>> deps;
+    public Optional<ImmutableSet<String>> contacts;
+    public Optional<ImmutableSet<Label>> labels;
+    @Hint(isDep = false) public Optional<ImmutableSortedSet<BuildTarget>> sourceUnderTest;
```

Comment:
Don't add `sourceUnderTest`, instead add `tests` to `go_library` and `go_binary`.